src/librustc_resolve/lib.rs

```diff
@@ -1663,16 +1663,16 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
         DefCollector::new(&mut definitions, Mark::root())
             .collect_root(crate_name, session.local_crate_disambiguator());
 
-        let mut extern_prelude: FxHashSet<Name> =
-            session.opts.externs.iter().map(|kv| Symbol::intern(kv.0)).collect();
+        let mut extern_prelude: FxHashMap<Ident, Option<&NameBinding>> =
+            session.opts.externs.iter().map(|kv| (Ident::from_str(kv.0), None)).collect();
 
-        // HACK(eddyb) this ignore the `no_{core,std}` attributes.
-        // FIXME(eddyb) warn (elsewhere) if core/std is used with `no_{core,std}`.
-        // if !attr::contains_name(&krate.attrs, "no_core") {
-        // if !attr::contains_name(&krate.attrs, "no_std") {
-        extern_prelude.insert(Symbol::intern("core"));
-        extern_prelude.insert(Symbol::intern("std"));
-        extern_prelude.insert(Symbol::intern("meta"));
+        if session.rust_2018() && !attr::contains_name(&krate.attrs, "no_core") {
+            if !attr::contains_name(&krate.attrs, "no_std") {
+                extern_prelude.insert(Ident::from_str("std"), None);
+            } else {
+                extern_prelude.insert(Ident::from_str("core"), None);
```

Comment:
I really preferred having `core` accessible even in non-`no_std` builds. I think the rest of these changes, but with this still outside this `else` gives the best pattern for crates with a `std` feature:

```rust
#![cfg_attr(not(feature = "std"), no_std)]

use core::mem;
#[cfg(feature = "std")]
use std::boxed::Box;
```

You get the best of both worlds, minimal overhead compared to a normal `std` using crate, plus compile-time checking that you don’t accidentally reference `std` in a `no_std` build (even when your target has `std` available).

Reply:
(If it’s possible there could be a warning about using `core` if your crate has no `cfg`-d out `no_std` attribute to push people that aren’t using this pattern to use `std` always, in case they copy-paste some `core` using code)

Reply:
I think using `core` instead of `std` where you don't have `#![no_std]` is nothing bad we should lint against; but using `std` when `#![no_std]` is active is a bad thing that we should (are?) linting with warn-by-default (or stronger) against. However, I agree with you that `use std::` should be possible when `#![no_std]` is active.

Reply:
> I agree with you that `use std::` should be possible when `#![no_std]` is active.

This is the current nightly’s behaviour. I think this should _not_ be possible, if it is ergonomic to do so. The only way to currently check whether you accidentally used `std` is to setup a build for a target like `thumbv6m-none-eabi` which doesn’t distribute a `std` lib, this is a real pain point. I think with the 2018 module changes, and this one change to always have `core` available (except `no_core`), it becomes ergonomic enough to have `#![no_std] use std::*;` be a hard error, and require crates to optionally enable `no_std` when they want to use `std` with a feature flag.

Reply:
> I agree with you that use std:: should be possible when #![no_std] is active.

I removed this bit when I realized it was wrong ;)

Reply:
>I really preferred having core accessible even in non-no_std builds.

Yeah, this is an uncontroversial addition.
What I want to avoid is anything that can't be done through the standard library prelude eventually.

Reply:
> What I want to avoid is anything that can't be done through the standard library prelude eventually.

Can you elaborate on why having `core` always available would lead to that?

Reply:
@Centril 
It's mostly about `std`, not `core` (`#[no_core]` is unstable anyway).

`core` can bring `core` into prelude when injected (`no_std && !no_core` scenario).
`std` can bring `std` and `core` into prelude when injected (`!no_std` scenario).

\*Nothing\* cannot bring anything into prelude (`no_core` scenario).
`core` cannot bring `std` into prelude when injected due to circular dependency (`no_std && !no_core` scenario).

Reply:
@petrochenkov Sorry; I don't follow what you are trying to say here...

I think the situation should be like so:

```rust
flag_std = true;
flag_core = true;

#![no_core] => [flag_core = false,  #![no_std]]
#![no_std] => [flag_std = false]

flag_core => core is in the prelude.
flag_std  => std is in the prelude.
```

Reply:
@Centril 
That's exactly what I was trying to say.

Reply:
@petrochenkov oh, ok, great :) let's do that.

Reply:
Not unless you add `extern crate std;` explicitly.
Starting with https://github.com/rust-lang/rust/pull/54116 (the first point) absolute paths only look into crates listed in extern prelude, not crates lying around in library search directories.